### PR TITLE
addDaysNoOverflow added

### PR DESCRIPTION
**Pull request by Maxim-Mazurok**
Inspired by [question](https://stackoverflow.com/q/51295911/4536543) on SO, I've created function `addDaysNoOverflow()`, that acts pretty much like `addMonthsNoOverflow()`.
Hope that it's ready for merge, thanks.

### CODE REVIEW

**Comment by test82987** [comment]
please, rename it to setDayNoOverflow and replace this $this->modify((int)$value . ' day'); with $this->day((int)$value) or you'll have problem with months.

**Comment by Maxim-Mazurok** [comment]
Thanks, @test82987, I've renamed function and added tests. Actually, my version passes tests, while your is failing. It may be because `modify` function will increase month and that's why we need last day of **previous** month. While your version just changes day.

**Comment by Maxim-Mazurok** [comment]
@test82987 I've added function setDaysNoOverflow with tests, check it out.

File: tests/Carbon/AddMonthsTest.php [hunk 1]
```diff
@@ -39,6 +39,28 @@ public function providerTestAddMonthNoOverflow()
...
     }
 
+    public function providerTestAddDaysNoOverflow()
+    {
+        return array(
+            array(-2, 2016, 1, 29),
+            array(-1, 2016, 1, 30),
+            array(0, 2016, 1, 31),
+            array(1, 2016, 1, 31),
+            array(2, 2016, 1, 31),
```
Comment (kylekatarnls) [inline tests/Carbon/AddMonthsTest.php]: You won't get expected results with big gap `-40`, `40`, `-100`, `100`, etc.

File: tests/Carbon/AddMonthsTest.php [hunk 2]
```diff
@@ -39,6 +39,28 @@ public function providerTestAddMonthNoOverflow()
...
+    }
+
+    public function providerTestSetDaysNoOverflow()
+    {
+        return array(
+            array(0, 2016, 1, 31),
+            array(1, 2016, 1, 1),
+            array(2, 2016, 1, 2),
+            array(32, 2016, 1, 31),
+            array(100, 2016, 1, 31),
```
Comment (kylekatarnls) [inline tests/Carbon/AddMonthsTest.php]: You won't get expected results with negative values `-1`, `-40`, `-100`, etc.

**Comment by test82987** [comment]
let's admit addDaysNoOverflow is not needed, when setDaysNoOverflow - does not require `-sub` version. Half problems done. Next - negatives should not work at all if you set negative date with no overflow there should be an exception. Next- big numbers - my suggestion throw an same exception if the number is more than 31. Next - zero - should give the same exception in my opinion.

**Comment by kylekatarnls** [comment]
`$date->setDaysNoOverflow($day)` is:
```php
$date->day(min($day, $date->daysInMonth));
```

Should be pretty equivalent to your code. And to allow negative numbers:
```php
$date->day(max(0, min($day, $date->daysInMonth)));
```

**Comment by imbrish** [comment]
Why not just (with `sub` variant):
```php
public function addDaysNoOverflow($value)
{
    return $this->setDaysNoOverflow($this->day + $value);
}
```
And, as @kylekatarnls proposed:
```php
public function setDaysNoOverflow($value)
{
    return $this->day(max(0, min($value, $this->daysInMonth)));
}
```
I think this method could be useful, despite it doesn't solve the actual problem from Stack Overflow.

On a side note, I find the "NoOverflow" suffix confusing, ie it's not clear what won't overflow just by looking at `addMonthsNoOverflow` and `addDaysNoOverflow`. Maybe something like `incrementMonth`, `addDaysWithinMonth` could work better?

**Comment by kylekatarnls** [comment]
Yes, add/sub should call a unified setter to avoid duplicate code. And there it seems we could have all those methods as 1 line of code. So maybe not so necessary.

Then if we have such possibility for day not overflow month, we could have it for any couple of units:
- set/add/sub hour without overflowing day
- set/add/sub week/month/quarter without overflowing year
- set/add/sub milli/microseconds without overflowing second/minute/hour
- etc.

As you can see, it will quickly add a plethora of methods.

**Comment by imbrish** [comment]
You're right that this belongs to the business logic not a date library. Like Taylor often says in Laravel repo, "create a macro" 😄

**Comment by kylekatarnls** [comment]
Maybe we should rather have a single method for that (only compatible with Carbon 2):
```php
public function setNoOverflow($value, $valueUnit, $overflowUnit)
{
    $original = $this->copy();

    return $original->set($valueUnit, $value)
        ->min($original->endOf($overflowUnit))
        ->max($original->startOf($overflowUnit));
}
```
Then addNoOverflow/subNoOverflow calling it. Then here you can set/add/sub any unit not overflowing any unit having endOf/startOf compatible with.

**Comment by imbrish** [comment]
I think this may even be a better than having a bunch of magic `__call` methods. Which applies also in case of other functionalities. As far as I remember this is how moment.js goes about that.

**Comment by kylekatarnls** [comment]
Not really, moment.js is a bit bad about this, it just disabled overflow for month since 2.1.0.

So before this version you can't prevent overflow, after this version, you can't allow overflow.

And a priori, there is nothing for other units.

**Comment by kylekatarnls** [comment]
Fixed via method `addUnitNoOverflow`, `subUnitNoOverflow` and  `setUnitNoOverflow` https://github.com/briannesbitt/Carbon/releases/tag/2.0.0-beta.1

**Pull request closed without merging**